From 143401de68998e065e9f2a18cfc42c2ae9bc1da8 Mon Sep 17 00:00:00 2001
From: Christian Roessner <c@roessner.co>
Date: Thu, 20 Aug 2026 08:19:08 +0200
Subject: [PATCH v2 1/2] bounce: mark locally generated DSNs

Set one internal cleanup flag only when bounce(8) posts a null-sender delivery status notification. Ordinary Postfix submissions, non-bounce notices, and externally received messages leave the flag clear.
---
 src/bounce/bounce_notify_service.c | 10 +++----
 src/bounce/bounce_notify_verp.c    | 10 +++----
 src/bounce/bounce_one_service.c    | 10 +++----
 src/bounce/bounce_trace_service.c  | 12 ++++++--
 src/bounce/bounce_warn_service.c   | 10 +++----
 src/global/cleanup_strflags.c      |  1 +
 src/global/cleanup_user.h          |  1 +
 src/global/post_mail.c             | 46 ++++++++++++++++++++++++++----
 src/global/post_mail.h             |  1 +
 9 files changed, 73 insertions(+), 28 deletions(-)

diff --git a/src/bounce/bounce_notify_service.c b/src/bounce/bounce_notify_service.c
index c6004c0bf..c464fbc2c 100644
--- a/src/bounce/bounce_notify_service.c
+++ b/src/bounce/bounce_notify_service.c
@@ -220,11 +220,11 @@ int     bounce_notify_service(int flags, char *service, char *queue_name,
      * NOTIFY restrictions.
      */
     else {
-	if ((bounce = post_mail_fopen_nowait(NULL_SENDER, recipient,
-					     MAIL_SRC_MASK_BOUNCE,
-					     NULL_TRACE_FLAGS,
-					     sendopts,
-					     new_id)) != 0) {
+	if ((bounce = post_mail_fopen_dsn_nowait(NULL_SENDER, recipient,
+						 MAIL_SRC_MASK_BOUNCE,
+						 NULL_TRACE_FLAGS,
+						 sendopts,
+						 new_id)) != 0) {
 
 	    /*
 	     * Send the bounce message header, some boilerplate text that
diff --git a/src/bounce/bounce_notify_verp.c b/src/bounce/bounce_notify_verp.c
index e0844599e..a95e930e5 100644
--- a/src/bounce/bounce_notify_verp.c
+++ b/src/bounce/bounce_notify_verp.c
@@ -163,11 +163,11 @@ int     bounce_notify_verp(int flags, char *service, char *queue_name,
 	    bounce_status = 0;
 	} else {
 	    verp_sender(verp_buf, verp_delims, recipient, rcpt);
-	    if ((bounce = post_mail_fopen_nowait(NULL_SENDER, STR(verp_buf),
-						 MAIL_SRC_MASK_BOUNCE,
-						 NULL_TRACE_FLAGS,
-						 sendopts,
-						 new_id)) != 0) {
+	    if ((bounce = post_mail_fopen_dsn_nowait(NULL_SENDER, STR(verp_buf),
+						     MAIL_SRC_MASK_BOUNCE,
+						     NULL_TRACE_FLAGS,
+						     sendopts,
+						     new_id)) != 0) {
 
 		/*
 		 * Send the bounce message header, some boilerplate text that
diff --git a/src/bounce/bounce_one_service.c b/src/bounce/bounce_one_service.c
index 2524cbc75..aab0de930 100644
--- a/src/bounce/bounce_one_service.c
+++ b/src/bounce/bounce_one_service.c
@@ -186,11 +186,11 @@ int     bounce_one_service(int flags, char *queue_name, char *queue_id,
 	    && (rcpt->dsn_notify & DSN_NOTIFY_FAILURE) == 0) {
 	    bounce_status = 0;
 	} else {
-	    if ((bounce = post_mail_fopen_nowait(NULL_SENDER, orig_sender,
-						 MAIL_SRC_MASK_BOUNCE,
-						 NULL_TRACE_FLAGS,
-						 sendopts,
-						 new_id)) != 0) {
+	    if ((bounce = post_mail_fopen_dsn_nowait(NULL_SENDER, orig_sender,
+						     MAIL_SRC_MASK_BOUNCE,
+						     NULL_TRACE_FLAGS,
+						     sendopts,
+						     new_id)) != 0) {
 
 		/*
 		 * Send the bounce message header, some boilerplate text that
diff --git a/src/bounce/bounce_trace_service.c b/src/bounce/bounce_trace_service.c
index e8c09cfa1..5d400fec9 100644
--- a/src/bounce/bounce_trace_service.c
+++ b/src/bounce/bounce_trace_service.c
@@ -176,11 +176,17 @@ int     bounce_trace_service(int flags, char *service, char *queue_name,
      * a new queue file.
      */
     new_id = vstring_alloc(10);
-    if ((bounce = post_mail_fopen_nowait(sender, recipient,
+    if (*sender == 0)
+	bounce = post_mail_fopen_dsn_nowait(sender, recipient,
+					     MAIL_SRC_MASK_BOUNCE,
+					     NULL_TRACE_FLAGS,
+					     sendopts, new_id);
+    else
+	bounce = post_mail_fopen_nowait(sender, recipient,
 					 MAIL_SRC_MASK_BOUNCE,
 					 NULL_TRACE_FLAGS,
-					 sendopts,
-					 new_id)) != 0) {
+					 sendopts, new_id);
+    if (bounce != 0) {
 	msg_info("%s: sender delivery status notification: %s",
 		 queue_id, STR(new_id));
 	count = -1;
diff --git a/src/bounce/bounce_warn_service.c b/src/bounce/bounce_warn_service.c
index 84ea22278..ddc457a23 100644
--- a/src/bounce/bounce_warn_service.c
+++ b/src/bounce/bounce_warn_service.c
@@ -208,11 +208,11 @@ int     bounce_warn_service(int unused_flags, char *service, char *queue_name,
      * restrictions.
      */
     else {
-	if ((bounce = post_mail_fopen_nowait(NULL_SENDER, recipient,
-					     MAIL_SRC_MASK_BOUNCE,
-					     NULL_TRACE_FLAGS,
-					     sendopts,
-					     new_id)) != 0) {
+	if ((bounce = post_mail_fopen_dsn_nowait(NULL_SENDER, recipient,
+						 MAIL_SRC_MASK_BOUNCE,
+						 NULL_TRACE_FLAGS,
+						 sendopts,
+						 new_id)) != 0) {
 
 	    /*
 	     * Send the bounce message header, some boilerplate text that
diff --git a/src/global/cleanup_strflags.c b/src/global/cleanup_strflags.c
index afcf47317..d4d9c95ea 100644
--- a/src/global/cleanup_strflags.c
+++ b/src/global/cleanup_strflags.c
@@ -56,6 +56,7 @@ static struct cleanup_flag_map cleanup_flag_map[] = {
     CLEANUP_FLAG_SMTPUTF8, "smtputf8_requested",
     CLEANUP_FLAG_AUTOUTF8, "smtputf8_autodetect",
     CLEANUP_FLAG_REQTLS, "requiretls_requested",
+    CLEANUP_FLAG_DSN_ORIGIN, "local_dsn_origin",
 };
 
 /* cleanup_strflags - map flags code to printable string */
diff --git a/src/global/cleanup_user.h b/src/global/cleanup_user.h
index 6d7c03ada..4ed9e3ca2 100644
--- a/src/global/cleanup_user.h
+++ b/src/global/cleanup_user.h
@@ -26,6 +26,7 @@
 #define CLEANUP_FLAG_SMTPUTF8	(1<<8)	/* SMTPUTF8 requested */
 #define CLEANUP_FLAG_AUTOUTF8	(1<<9)	/* Autodetect SMTPUTF8 */
 #define CLEANUP_FLAG_REQTLS	(1<<10)	/* REQUIRETLS requested */
+#define CLEANUP_FLAG_DSN_ORIGIN	(1<<11)	/* Local bounce(8) DSN */
 
 #define CLEANUP_FLAG_FILTER_ALL	(CLEANUP_FLAG_FILTER | CLEANUP_FLAG_MILTER)
 
diff --git a/src/global/post_mail.c b/src/global/post_mail.c
index f787acc8f..5e41c2c26 100644
--- a/src/global/post_mail.c
+++ b/src/global/post_mail.c
@@ -24,6 +24,15 @@
 /*	int	sendopts;
 /*	VSTRING *queue_id;
 /*
+/*	VSTREAM	*post_mail_fopen_dsn_nowait(sender, recipient, source_class,
+/*					trace_flags, sendopts, queue_id)
+/*	const char *sender;
+/*	const char *recipient;
+/*	int	source_class;
+/*	int	trace_flags;
+/*	int	sendopts;
+/*	VSTRING *queue_id;
+/*
 /*	void	post_mail_fopen_async(sender, recipient, source_class,
 /*					trace_flags, sendopts,
 /*					queue_id, notify, context)
@@ -75,6 +84,11 @@
 /*	only once, and does not wait until the cleanup service is
 /*	available.  Otherwise it is identical to post_mail_fopen().
 /*
+/*	post_mail_fopen_dsn_nowait() is a bounce(8)-specific interface
+/*	for posting a locally-generated delivery status notification.
+/*	The sender must be the null address and the source class must
+/*	be MAIL_SRC_MASK_BOUNCE.
+/*
 /*	post_mail_fopen_async() contacts the cleanup service and
 /*	invokes the caller-specified notify routine, with the
 /*	open stream and the caller-specified context when the
@@ -218,7 +232,8 @@ typedef struct {
 static void post_mail_init(VSTREAM *stream, const char *sender,
 			           const char *recipient,
 			           int source_class, int trace_flags,
-			           int sendopts, VSTRING *queue_id)
+			           int sendopts, int dsn_origin,
+			           VSTRING *queue_id)
 {
     VSTRING *id = queue_id ? queue_id : vstring_alloc(100);
     struct timeval now;
@@ -227,7 +242,8 @@ static void post_mail_init(VSTREAM *stream, const char *sender,
     int_filt_flags(source_class) | CLEANUP_FLAG_MASK_INTERNAL
     | smtputf8_autodetect(source_class)
     | ((sendopts & SOPT_REQUIRETLS_ESMTP) ? CLEANUP_FLAG_REQTLS : 0)
-    | ((sendopts & SMTPUTF8_FLAG_REQUESTED) ? CLEANUP_FLAG_SMTPUTF8 : 0);
+    | ((sendopts & SMTPUTF8_FLAG_REQUESTED) ? CLEANUP_FLAG_SMTPUTF8 : 0)
+    | (dsn_origin ? CLEANUP_FLAG_DSN_ORIGIN : 0);
 
     GETTIMEOFDAY(&now);
     date = mail_date(now.tv_sec);
@@ -297,7 +313,7 @@ VSTREAM *post_mail_fopen(const char *sender, const char *recipient,
 
     stream = mail_connect_wait(MAIL_CLASS_PUBLIC, var_cleanup_service);
     post_mail_init(stream, sender, recipient, source_class, trace_flags,
-		   sendopts, queue_id);
+		   sendopts, 0, queue_id);
     return (stream);
 }
 
@@ -312,7 +328,27 @@ VSTREAM *post_mail_fopen_nowait(const char *sender, const char *recipient,
     if ((stream = mail_connect(MAIL_CLASS_PUBLIC, var_cleanup_service,
 			       BLOCKING)) != 0)
 	post_mail_init(stream, sender, recipient, source_class, trace_flags,
-		       sendopts, queue_id);
+		       sendopts, 0, queue_id);
+    else
+	msg_warn("connect to %s/%s: %m",
+		 MAIL_CLASS_PUBLIC, var_cleanup_service);
+    return (stream);
+}
+
+/* post_mail_fopen_dsn_nowait - post a local bounce(8) DSN */
+
+VSTREAM *post_mail_fopen_dsn_nowait(const char *sender, const char *recipient,
+				            int source_class, int trace_flags,
+				            int sendopts, VSTRING *queue_id)
+{
+    VSTREAM *stream;
+
+    if (*sender != 0 || source_class != MAIL_SRC_MASK_BOUNCE)
+	msg_panic("post_mail_fopen_dsn_nowait: invalid argument");
+    if ((stream = mail_connect(MAIL_CLASS_PUBLIC, var_cleanup_service,
+			       BLOCKING)) != 0)
+	post_mail_init(stream, sender, recipient, source_class, trace_flags,
+		       sendopts, 1, queue_id);
     else
 	msg_warn("connect to %s/%s: %m",
 		 MAIL_CLASS_PUBLIC, var_cleanup_service);
@@ -341,7 +377,7 @@ static void post_mail_open_event(int event, void *context)
 	non_blocking(vstream_fileno(state->stream), BLOCKING);
 	post_mail_init(state->stream, state->sender,
 		       state->recipient, state->source_class,
-		       state->trace_flags, state->sendopts,
+		       state->trace_flags, state->sendopts, 0,
 		       state->queue_id);
 	myfree(state->sender);
 	myfree(state->recipient);
diff --git a/src/global/post_mail.h b/src/global/post_mail.h
index 2e855c3d0..37006230e 100644
--- a/src/global/post_mail.h
+++ b/src/global/post_mail.h
@@ -31,6 +31,7 @@
 typedef void (*POST_MAIL_NOTIFY) (VSTREAM *, void *);
 extern VSTREAM *post_mail_fopen(const char *, const char *, int, int, int, VSTRING *);
 extern VSTREAM *post_mail_fopen_nowait(const char *, const char *, int, int, int, VSTRING *);
+extern VSTREAM *post_mail_fopen_dsn_nowait(const char *, const char *, int, int, int, VSTRING *);
 extern void post_mail_fopen_async(const char *, const char *, int, int, int, VSTRING *, POST_MAIL_NOTIFY, void *);
 extern int PRINTFLIKE(2, 3) post_mail_fprintf(VSTREAM *, const char *,...);
 extern int post_mail_fputs(VSTREAM *, const char *);
-- 
2.55.0

